### PR TITLE
fix: gc to continue on retrieval error

### PR DESCRIPTION
**Pull request by istae**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### Screenshots (if appropriate):

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3585)

### CODE REVIEW

File: pkg/localstore/mode_put.go [hunk 1]
```diff
@@ -466,6 +466,11 @@ func (db *DB) addToCache(
 	batch *leveldb.Batch,
 	item shed.Item,
 ) (int64, error) {
+
+	err := db.pullIndex.DeleteInBatch(batch, item)
```
Comment (aloknerurkar) [inline pkg/localstore/mode_put.go:470]: So moving these doesnt make any difference. The index is only updated when the Batch is written. So order doesnt matter.

File: pkg/localstore/mode_set.go [hunk 2]
```diff
@@ -365,6 +365,10 @@ func (db *DB) setUnpin(batch *leveldb.Batch, addr swarm.Address) (gcSizeChange i
 	item.StoreTimestamp = i.StoreTimestamp
 	item.BinID = i.BinID
 	item.BatchID = i.BatchID
+	err = db.pullIndex.DeleteInBatch(batch, item)
```
Comment (aloknerurkar) [inline pkg/localstore/mode_set.go:368]: Same

File: pkg/localstore/gc.go [hunk 3]
```diff
@@ -179,6 +179,9 @@ func (db *DB) collectGarbage() (evicted uint64, done bool, err error) {
 
 		storedItem, err := db.retrievalDataIndex.Get(item)
 		if err != nil {
+			if errors.Is(err, leveldb.ErrNotFound) {
```
Comment (aloknerurkar) [inline pkg/localstore/gc.go:182]: So I had thought about this. But ideally this shouldnt happen as gcIndex is also deleted once it is removed.

Either way, if retrievalDataIndex is not present, then we should atleast deleted the gcIndex so we dont go through this entry again.

File: pkg/localstore/mode_set.go (hunk 2 above)
Comment (istae) [reply pkg/localstore/mode_set.go:368]: so here if an item is in the pushIndex, it should not affect the reserve, this is why it's moved above the pushIndex check.

File: pkg/localstore/mode_put.go (hunk 1 above)
Comment (istae) [reply pkg/localstore/mode_put.go:470]: Here we have early returns from the function due to the pin check, this should not affect the chunk being removed from the reserve.

File: pkg/localstore/gc.go (hunk 3 above)
Comment (istae) [reply pkg/localstore/gc.go:182]: yes should not happen but we have thousands of "msg"="collect garbage failed" errors

File: pkg/localstore/mode_put.go (hunk 1 above)
Comment (aloknerurkar) [reply pkg/localstore/mode_put.go:470]: So on L481 we return if there is an existing pinIndex. In this case, we dont put the item to gcIndex and hence we should not delete the pullIndex.

**Review by aloknerurkar — APPROVED** [review]

**Pull request merged**